common/src/ess.rs

```diff
@@ -0,0 +1,189 @@
+    }
+}
+
+#[async_trait]
+impl<T> ChannelService for SharedEss<T>
+where
+    T: Clone + Send + Sync + 'static,
+{
+    type OpenStream = ReceiverStream<Result<Event, Status>>;
+
+    async fn open(
+        &self,
+        _: tonic::Request<OpenRequest>,
+    ) -> Result<Response<Self::OpenStream>, Status> {
+        const METADATA_KEY: &str = "x-chariott-channel-id";
+
+        let id = Uuid::new_v4().to_string();
+        let (_, receiver_stream) = self.read_events(id.clone().into());
+        let mut response = Response::new(receiver_stream);
+        response.metadata_mut().insert(METADATA_KEY, id.try_into().unwrap());
+        Ok(response)
+    }
+}
+
+impl<T> Deref for SharedEss<T> {
+    type Target = InnerEss<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.as_ref()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
+
+    use crate::proto::{
+        common::Value as ValueMessage,
+        common::{value::Value as ValueEnum, SubscribeIntent},
+        streaming::{channel_service_server::ChannelService, OpenRequest},
+    };
+    use tokio_stream::StreamExt as _;
+    use tonic::{Code, Request};
+
+    use super::SharedEss;
+
+    #[tokio::test]
+    async fn open_should_set_channel_id() {
+        // arrange
+        let subject = setup();
+
+        // act
+        let response = subject.open(Request::new(OpenRequest {})).await.unwrap();
+
+        // assert
+        assert!(!response.metadata().get("x-chariott-channel-id").unwrap().is_empty());
+    }
+
+    #[tokio::test]
+    async fn serve_subscriptions_should_serve_subscription_for_event() {
+        // arrange
+        const EVENT_A: &str = "test-event-a";
+        const EVENT_B: &str = "test-event-b";
+
+        let subject = setup();
+        let response = subject.open(Request::new(OpenRequest {})).await.unwrap();
+        let channel_id =
+            response.metadata().get("x-chariott-channel-id").unwrap().to_str().unwrap().into();
+
+        // act
+        subject
+            .serve_subscriptions(
+                SubscribeIntent { channel_id, sources: vec![EVENT_A.into(), EVENT_B.into()] },
+                |_| ValueEnum::Null(0),
+            )
+            .unwrap();
+
+        // assert
+        subject.publish(EVENT_A, ());
+        subject.publish(EVENT_B, ());
+
+        let result = response
+            .into_inner()
+            .timeout(Duration::from_millis(100))
+            .take_while(|e| e.is_ok())
+            .map(|e| e.unwrap().unwrap())
+            .collect::<Vec<_>>()
+            .await;
+
+        // assert sources
+        assert_eq!(
+            vec![EVENT_A, EVENT_B],
+            result.iter().map(|e| e.source.clone()).collect::<Vec<_>>()
+        );
+
+        // assert sequence numbers
+        assert_eq!(1, result[0].seq);
+        assert_eq!(1, result[1].seq);
+
+        // assert payload
+        assert_eq!(Some(ValueMessage { value: Some(ValueEnum::Null(0)) }), result[0].value);
```

Comment:
nit: Instead of different assert blocks, this could just be a loop over expected sources and other data being constant?

```suggestion
        let expected_sources = [EVENT_A, EVENT_B];

        assert_eq!(expected_sources.len(), result.len());

        for (expected_source, actual) in expected_sources.iter().zip(result) {
            assert_eq!(*expected_source, actual.source.as_str());
            assert_eq!(1, actual.seq);
            assert_eq!(Some(ValueMessage { value: Some(ValueEnum::Null(0)) }), actual.value);
        }
```

Reply:
Thanks, applied your suggestion with a slight modification in bc8dbc96723ac6faa5869a57f355b95e881e8a66.